### PR TITLE
Replace VerifyWithRevocation test with DynamicRevocationTests

### LINKED ISSUE #41952: Test failure: System.Security.Cryptography.X509Certificates.Tests.ChainTests.VerifyWithRevocation
failed in job: [runtime-libraries-coreclr outerloop 20200907.3 ](https://dev.azure.com/dnceng/public/_build/results?buildId=804427&view=ms.vss-test-web.build-test-results-tab&runId=25478140&resultId=103383&paneView=debug)

net5.0-Linux-Release-arm64-CoreCLR_release-(Ubuntu.1804.ArmArch.Open)Ubuntu.1804.ArmArch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-16.04-helix-arm64v8-bfcd90a-20200127194925

Error message
~~~
Assert.Equal() Failure
Expected: 3
Actual:   2


Stack trace
   at System.Security.Cryptography.X509Certificates.Tests.ChainTests.VerifyWithRevocation() in /_/src/libraries/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:line 785
~~~

### PR DESCRIPTION
**Pull request by vcsjones**
`VerifyWithRevocation` is a flaky test. It needs internet to contact a CA for CRL/AIA and can occasionally fail. This replaces the offline revocation checking part of that test with `DynamicChainTests` which use our own local CA.

Fixes #41952

### CODE REVIEW

**Comment by vcsjones** [comment]
Some justification:

I ran `VerifyWithRevocation` on an ARM64 AWS instance a few thousand times, dumping the CRL and OCSP cache between each run by deleting `~/.dotnet/corefx/cryptography/crls` and `~/.dotnet/corefx/cryptography/ocsp`. I was unable to get it to fail with a stable network connection, so I am not aware of an actual underlying issue that the test is catching.

File: src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/DynamicRevocationTests.cs [hunk 1]
```diff
@@ -990,6 +990,95 @@ public static void TestRevocationWithNoNextUpdate_Revoked()
...
+                    offlineChain.ChainPolicy.RevocationMode = X509RevocationMode.Offline;
+
+                    SimpleRevocationBody(
+                        onlineHolder,
+                        endEntity,
+                        rootRevoked: false,
+                        issrRevoked: false,
+                        leafRevoked: false);
+
+                    SimpleRevocationBody(
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/DynamicRevocationTests.cs:1014]: Ideally we would shut down the responder between these two calls.

The easiest way is to probably do something like

* add

```C#
private delegate void RunSimpleTestWithResponder(
            RevocationResponder responder,
            CertificateAuthority root,
            CertificateAuthority intermediate,
            X509Certificate2 endEntity,
            ChainHolder chainHolder);
```

Then make there be three overloads of SimpleTest:

```C#
        private static void SimpleTest(
            PkiOptions pkiOptions,
            RunSimpleTest callback,
            [CallerMemberName] string callerName = null,
            bool pkiOptionsInTestName = true) =>
                SimpleTest(pkiOptions, callback, null, callerName, pkiOptionsInTestName);

        private static void SimpleTest(
            PkiOptions pkiOptions,
            RunSimpleTestWithResponder callback,
            [CallerMemberName] string callerName = null,
            bool pkiOptionsInTestName = true) =>
                SimpleTest(pkiOptions, null, callback, callerName, pkiOptionsInTestName);

        private static void SimpleTest(
            PkiOptions pkiOptions,
            RunSimpleTest? callback,
            RunSimpleTestWithResponder? callbackWithResponder,
            string callerName,
            bool pkiOptionsInTestName)
        {
            Assert.True((callback == null) != (callbackWithResponder == null), "Caller provided exactly one callback option");

            // existing body, adjusted to use whichever callback it got.
            ...
```

Comment (bartonjs) [reply src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/DynamicRevocationTests.cs:1017]: Editing every existing test to be able to adjust the delegate also works, of course :smile:.

**Review by bartonjs — APPROVED** [review]

**Comment by vcsjones** [comment]
This could use an outerloop run when anyone has a moment: `/azp run runtime-libraries-coreclr outerloop`

**Comment by bartonjs** [comment]
None of the failures in HTTP (which look to be various kinds of timeouts) should have been caused by this test-only change (at least, it's not obvious that the timeouts are caused by some sort of automagical call to RevocationResponder.Stop()), so merging.

**Comment by vcsjones** [comment]
@bartonjs I don't know what the bar is for porting test enhancements, but would it make sense to port this and perhaps #42680 to release/5.0 to stabilize the branch?

**Pull request merged**